### PR TITLE
Allow login via HPI OpenID

### PR DESCRIPTION
**Pull request by cmfcmf**
This is part of #637.

All HPI users are automatically promoted to coaches.
We still use Devise, which has support for OmniAuth.
The link to login is located here:
image

Further providers could be added, but I thought HPI OpenID was most important to begin with.

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -39,7 +39,10 @@
   devise_scope :user do
     get "/users/sign_up" => redirect("/users/sign_in")
   end
-  devise_for :users, :controllers => {:registrations => "users/registrations"}
+  devise_for :users, :controllers => {
+      :registrations => "users/registrations",
+      :omniauth_callbacks => "users/omniauth_callbacks"
```
Comment (bjrne) [inline config/routes.rb:44]: Can you explain what these lines are for? :)

**Review by bjrne — APPROVED** [review]
Looks good 👑

File: config/routes.rb (hunk 1 above)
Comment (cmfcmf) [reply config/routes.rb:44]: They tell Devise to use our controllers (inside `app/controllers`) for `:registrations` and `:omniauth_callbacks`.

File: Gemfile [hunk 2]
```diff
@@ -35,7 +35,9 @@ gem 'jquery-ui-rails'
 # Authentication
 gem 'devise'
 # openID Authentication
-gem 'devise_openid_authenticatable'
+#gem 'devise_openid_authenticatable'
```
Comment (corinnaj) [inline Gemfile:38]: This comment can probably be deleted

File: Gemfile.lock [hunk 3]
```diff
@@ -71,7 +71,7 @@ GEM
       debug_inspector (>= 0.0.1)
     bootstrap-datepicker-rails (1.6.4.1)
       railties (>= 3.0)
-    builder (3.2.2)
+    builder (3.2.3)
```
Comment (corinnaj) [inline Gemfile.lock:74]: Out of curiosity, why are there so many version changes?

File: app/models/user.rb [hunk 4]
```diff
@@ -42,6 +45,19 @@ def set_default_role
     self.role ||= :pupil
   end
 
+  def self.from_omniauth(auth)
+    where(provider: auth.provider, uid: auth.uid).first_or_create! do |user|
+      user.email = auth.info.email
+      user.password = Devise.friendly_token[0,20]
+      user.role = :coach
+      # user.name = auth.info.first_name + " " + auth.info.last_name
```
Comment (corinnaj) [inline app/models/user.rb:53]: Another comment, that is probably not needed

File: spec/features/user_hpiopenid_spec.rb [hunk 5]
```diff
@@ -0,0 +1,45 @@
+require "rails_helper"
+
+# https://www.relishapp.com/rspec/rspec-rails/v/3-5/docs/feature-specs/feature-spec
+RSpec.feature "Account creation vis HPI OpenID", :type => :feature do
```
Comment (corinnaj) [inline spec/features/user_hpiopenid_spec.rb:4]: There's a typo here: vis should probably be via

File: spec/views/users/users_sessions_new.html.erb_spec.rb [hunk 6]
```diff
@@ -1,12 +1,17 @@
 require 'rails_helper'
 
 
-RSpec.describe "users/sessions/new", type: :view do 
-  
+RSpec.describe "users/sessions/new", type: :view do
+  before(:each) do
+    without_partial_double_verification do
```
Comment (corinnaj) [inline spec/views/users/users_sessions_new.html.erb_spec.rb:6]: Can you explain to me what lines 6 and 7 do?

**Review by corinnaj — APPROVED** [review]
Looks good to me, I only found some very minor things and a few questions.

**Review by corinnaj — CHANGES_REQUESTED** [review]
I get an error when I try to create a new account with OpenID on my machine: 
```ActiveRecord::RecordInvalid - Gültigkeitsprüfung ist fehlgeschlagen: E-Mail-Adresse Feld ist leer: bitte fülle es aus, um fortzufahren.```

**Comment by bjrne** [comment]
That's weird because it works on heroku. Did you set up your openID Account correctly?

**Comment by corinnaj** [comment]
For me it doesn't work on Heroku either. We tried it with @aloeser account as well and got the same error:
screenshot from 2017-06-19 10-34-07

**Comment by bjrne** [comment]
Okay, since this error was tied to a wrong configuration on the openID provider site I'd suggest to merge this for now and make a separate issue (#664) for an explanation along with a proper error message @corinnaj @aloeser ?

**Review by corinnaj — APPROVED** [review]

**Comment by bjrne** [comment]
```
2017-06-20T12:15:58.220488+00:00 app[web.1]: Completed 500 Internal Server Error in 4ms (ActiveRecord: 1.0ms)
2017-06-20T12:15:58.232738+00:00 app[web.1]: 
2017-06-20T12:15:58.232740+00:00 app[web.1]: ActiveRecord::StatementInvalid (PG::UndefinedColumn: ERROR:  column users.provider does not exist
2017-06-20T12:15:58.232740+00:00 app[web.1]: LINE 1: SELECT  "users".* FROM "users" WHERE "users"."provider" = 'h...
2017-06-20T12:15:58.232741+00:00 app[web.1]:                                              ^
2017-06-20T12:15:58.232742+00:00 app[web.1]: : SELECT  "users".* FROM "users" WHERE "users"."provider" = 'hpiopenid' AND "users"."uid" = 'https://openid.hpi.uni-potsdam.de/user/bjarne.sievers'  ORDER BY "users"."id" ASC LIMIT 1):
2017-06-20T12:15:58.232742+00:00 app[web.1]:   app/models/user.rb:49:in `from_omniauth'
2017-06-20T12:15:58.232743+00:00 app[web.1]:   app/controllers/users/omniauth_callbacks_controller.rb:5:in `hpiopenid'
2017-06-20T12:15:58.232743+00:00 app[web.1]: 
2017-06-20T12:15:58.232744+00:00 app[web.1]: 
2017-06-20T12:15:58.233226+00:00 heroku[router]: at=info method=GET path="/users/auth/hpiopenid/callback?_method=post&openid.ns=http%3A%2F%2Fspecs.openid.net%2Fauth%2F2.0&openid.op_endpoint=https%3A%2F%2...
```

**Pull request merged**